File: go/blackjack/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) (value int) {
	switch card {
	case "ace":
		value = 11
	case "two":
		value = 2
	case "three":
		value = 3
	case "four":
		value = 4
	case "five":
		value = 5
	case "six":
		value = 6
	case "seven":
		value = 7
	case "eight":
		value = 8
	case "nine":
		value = 9
	case "king", "queen", "jack", "ten":
		value = 10
	}
	return
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	var (
		// parse cards
		c1, c2, dc = ParseCard(card1), ParseCard(card2), ParseCard(dealerCard)
		// establish hand
		hand = c1 + c2
		// determine outcome modifiers
		split, sevens_high, dealer_match = (c1 == c2), (dc >= 7), (c1 == dc || c2 == dc)
	)
	switch {
	case (hand > 21 && split) || (hand >= 12 && !sevens_high && split):
		return "P" // Split (P)
	case (hand == 21 && dealer_match) || (hand == 20) || (hand < 21 && hand >= 17) || (hand < 21 && hand >= 12 && !sevens_high && !split):
		return "S" // Stand (S)
	case (hand == 21 && !dealer_match):
		return "W" // Automatically win (W)
	case (hand >= 12 && sevens_high) || (hand <= 11):
		return "H" // Hit (H)
	default:
		return "unknown"
	}
}

File: ruby/blackjack/blackjack.rb
module Blackjack
  def self.parse_card(card)
    ["joker", "one", "two", "three", "four", "five",
    "six", "seven", "eight", "nine", "ten","ace"].index(card) || 10 # jack, queen and king
  end

  def self.card_range(card1, card2)
    low = (4..11)
    mid = (12..16)
    high = (17..20)
    bj = 21
    case parse_card(card1) + parse_card(card2)
    when low then 'low'
    when mid then 'mid'
    when high then 'high'
    when bj then 'blackjack'
    end
  end

  def self.first_turn(card1, card2, dealer_card)
    c1 = parse_card(card1)
    c2 = parse_card(card2)
    d1 = parse_card(dealer_card)
    cr = card_range(card1, card2)
    case
    when (c1 == c2 && c1 == 11) then "P" 
    when cr == "blackjack" then (10..11) === d1  ? "S" : "W"
    when cr == "high" then "S"
    when cr == "mid" then d1 >= 7 ?  "H" : "S" 
    when cr == "low" then "H"
    end
  end
end
